src/meshapi/util/uisp_import/sync_handlers.py

```diff
@@ -0,0 +1,297 @@
+                f"During UISP link import link with UISP ID {uisp_uuid} was skipped "
+                f"because the data in UISP references a 'from' device (UISP ID {uisp_from_device_uuid}) "
+                f"which we do not have in our database (perhaps it was skipped at device import time?)"
+            )
+            continue
+
+        if not uisp_link["to"]["device"]:
+            logging.warning(
+                f"During UISP link import link with UISP ID {uisp_uuid} was skipped "
+                f"because the data in UISP does not indicate what device this link connects to"
+            )
+            continue
+
+        uisp_to_device_uuid = uisp_link["to"]["device"]["identification"]["id"]
+        try:
+            uisp_to_device = Device.objects.get(uisp_id=uisp_to_device_uuid)
+        except Device.DoesNotExist:
+            logging.warning(
+                f"During UISP link import link with UISP ID {uisp_uuid} was skipped "
+                f"because the data in UISP references a 'to' device (UISP ID {uisp_to_device_uuid}) "
+                f"which we do not have in our database (perhaps it was skipped at device import time?)"
+            )
+            continue
+
+        if uisp_link["state"] == "active":
+            uisp_status = Link.LinkStatus.ACTIVE
+        else:
+            uisp_status = Link.LinkStatus.INACTIVE
+
+        uisp_link_type = get_link_type(uisp_link)
+
+        with transaction.atomic():
+            existing_links: List[Link] = list(Link.objects.filter(uisp_id=uisp_uuid).select_for_update())
+            if existing_links:
+                if len(existing_links) > 1:
+                    notify_administrators_of_data_issue(
+                        existing_links,
+                        LinkSerializer,
+                        message=f"Possible duplicate objects detected, links share the same UISP ID ({uisp_uuid})",
+                    )
+
+                for existing_link in existing_links:
+                    change_list = update_link_from_uisp_data(
+                        existing_link,
+                        uisp_from_device,
+                        uisp_to_device,
+                        uisp_status,
+                        uisp_link_type,
+                        uisp_session,
+                    )
+                    if change_list:
+                        notify_admins_of_changes(existing_link, change_list)
+                continue
+
+        link = Link(
+            from_device=uisp_from_device,
+            to_device=uisp_to_device,
+            status=uisp_status,
+            type=uisp_link_type,
+            uisp_id=uisp_uuid,
+            notes=f"Automatically imported from UISP on {datetime.date.today().isoformat()}\n\n",
+            # UISP doesn't track the following info
+            install_date=None,
+            abandon_date=None,
+            description=None,
+        )
+        link.save()
+
+
+def sync_link_table_into_los_objects() -> None:
```

Comment:
Oh man it kinda sucks that we have to do this. I wonder if a Link could have a mandatory LOS field? Maybe that wouldn't work for things like fiber...

Reply:
Strictly speaking we don't have to do this, it's just going to make the LOS table more useful to do it. For the map to render accurately we don't need any LOS objects that duplicate links. 

Regarding some kind of foreign key with link, I think that could possibly be useful, but I think it would have to be m2m and optional (fiber is a great example of this as you point out). It feels like for our current needs it is overkill and will be more trouble than it's worth, but perhaps that will change as we work towards Map v2 TM (and probably LOS features under the same umbrella).

Good call about fiber though! We should definitely exclude fiber and ethernet links from our assumption that "every link implies a LOS"

Reply:
Also VPN 🫣